src/snapshot_tests/test_case.rs

```diff
@@ -0,0 +1,277 @@
+            let output_frame = outputs.frames.get(&spec.output_id).unwrap();
+            let new_snapshot = frame_to_rgba(output_frame);
+            snapshots.push(Snapshot {
+                test_name: self.name.to_owned(),
+                output_id: spec.output_id.clone(),
+                pts,
+                resolution: output_frame.resolution,
+                data: new_snapshot,
+            });
+        }
+
+        Ok(snapshots)
+    }
+
+    fn check_for_unused_snapshots(&self, snapshots: &[Snapshot]) -> Result<()> {
+        let snapshot_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
+            .join(SNAPSHOTS_DIR_NAME)
+            .join(self.name);
+        let parent_path = snapshot_path.parent().unwrap();
+
+        for entry in fs::read_dir(parent_path).unwrap() {
+            let entry = entry.unwrap();
+            if !entry.file_name().to_string_lossy().ends_with(".bmp") {
+                continue;
+            }
+            if !snapshots.iter().any(|s| s.save_path() == entry.path()) {
+                return Err(TestCaseError::UnusedSnapshot(entry.path()).into());
+            }
+        }
+        Ok(())
+    }
+
+    pub fn prepare_renderer_and_scene(&self) -> (Renderer, Arc<SceneSpec>) {
+        fn register_requests_to_renderers(register_request: RegisterRequest) -> RendererSpec {
+            match register_request {
+                RegisterRequest::InputStream(_) | RegisterRequest::OutputStream(_) => {
+                    panic!("Input and output streams are not supported in snapshot tests")
+                }
+                RegisterRequest::Shader(shader) => shader.try_into().unwrap(),
+                RegisterRequest::WebRenderer(web_renderer) => web_renderer.try_into().unwrap(),
+                RegisterRequest::Image(img) => img.try_into().unwrap(),
+            }
+        }
+
+        if self.name.is_empty() {
+            panic!("Snapshot test name has to be provided");
+        }
+        let renderers: Vec<RendererSpec> = self
+            .renderers
+            .iter()
+            .cloned()
+            .map(|json| serde_json::from_str(json).unwrap())
+            .map(register_requests_to_renderers)
+            .collect();
+
+        let scene: Scene = serde_json::from_str(self.scene_json).unwrap();
+        let scene: Arc<SceneSpec> = Arc::new(scene.try_into().unwrap());
+
+        let renderer = create_renderer(renderers, scene.clone());
+        (renderer, scene)
+    }
+}
+
+#[derive(Debug)]
+pub struct TestInput {
+    pub name: String,
+    pub resolution: Resolution,
+    pub data: YuvData,
+}
+
+impl TestInput {
+    pub fn new(index: usize) -> Self {
+        Self::new_with_resolution(
+            index,
+            Resolution {
+                width: 640,
+                height: 360,
+            },
+        )
+    }
+
+    pub fn new_with_resolution(index: usize, resolution: Resolution) -> Self {
+        let color = ((index * 123) % 255) as u8;
```

Comment:
is there a way to easily identify which color is which? If not maybe consider just hardcoding list of colors. Up to index_16 should be enough, but make sure first few are the most distinctive ones e.g. red green and blue